content/post/2019-12-19-mutable-api.Rmarkdown

```diff
@@ -0,0 +1,184 @@
+
+In the `urltools` package there are a few functions for getting or setting parts of an URL such as the fragment.
+
+```{r}
+url <- "https://docs.r-hub.io/#package-builder"
+urltools::fragment(url)
+urltools::fragment(url) <- "intro"
+url
+urltools::fragment(url) <- NULL
+url
+```
+
+The original url value is not modified, the url name is bound to a new binding. Below we use the [`tracemem()` function](https://adv-r.hadley.nz/names-values.html#tracemem).
+
+```{r}
+url <- "https://docs.r-hub.io/#package-builder"
+tracemem(url)
+urltools::fragment(url) <- "intro"
+url
+
+```
+
+So how does the above work, exactly? What's that `fragment` method?
+
+```{r}
+getMethod(urltools::"fragment<-")
+```
+
+Actually, reading the source above doesn't help us. Sure it creates a new string, but how on Earth is it the new string bound to the initial name? Well, it's because the function is called `fragment<-` with an arrow at the end and has a last argument called `value`, both criteria together make it a **replacement function**. Replacement functions are presented in [The R Language definition](https://cran.r-project.org/doc/manuals/R-lang.html#Subset-assignment), [this StackOverflow thread](https://stackoverflow.com/questions/11563154/what-are-replacement-functions-in-r) and [in the Advanced R book by Hadley Wickham](https://adv-r.hadley.nz/functions.html#replacement-functions).
+
+Let's create our own replacement function to make sure we got it right!
+
+```{r}
+x <- 1:5
+x
+
+# function that will replace all values of x
+# with the new value
+`replace_all<-` <- function(x, value) {
+  x[seq_along(x)] <- value
+  x
+}
+
+# the argument called value is passed at the right of the arrow
+replace_all(x) <- 42
+x
+```
+
+So we've modified x, but not in place, see below the same code with `tracemem()`
+
+```{r}
+x <- 1:5
+tracemem(x)
+`replace_all<-` <- function(x, value) {
+  x[seq_along(x)] <- value
+  x
+}
+replace_all(x) <- 42
+x
+```
+
+So replacement functions are a standard way to give a mutable flavour to R code. Let's move on to another mutable feel.
+
+## Exposing the C API in xml2
+
+With `xml2` you can modify and remove XML nodes from a tree which makes you feel the tree is mutable.
+
+See for instance the code in [our blog post about READMEs](/2019/12/03/readmes/#other-size-indicators)
+
+```r
+xml2::xml_replace(xml2::xml_find_all(xml, "//softbreak"),
+                      xml2::read_xml("<text>\n</text>"))
+```
+
+That code changes nodes in the `xml` object without our assigning it back to it.
+
+The reasons `xml2` works this way is that the package is a binding to [the C libxml2 API](http://xmlsoft.org/), where you are supposed to manage memory allocation manually. [`xml2` does handle memory management for you](https://xml2.r-lib.org/#compared-to-the-xml-package) but [`xml2` docs explain how one should be careful when removing nodes because of possible memory issues](https://xml2.r-lib.org/articles/modification.html#removing-nodes). 
```

Comment:
> where you are supposed to manage memory allocation manually.

We could just replace this with "and the XML objects in the libxml2 C API _are_ mutable."

> how one should be careful when removing nodes because of possible memory issues

I would leave this out, not relevant here I think. People might think that this is typical, and it is not.